### PR TITLE
Implement MSIDKeyedArchiverSerializer

### PR DESCRIPTION
**Pull request by antrix1989**
Final version. Added new properties to MSIDToken:
- resource
- authority
- clientId
- scopes

Ignore *accessTokenType* when map to/from MSIDToken.  Seems that *accessTokenType* always has "Bearer" type, moreover, we parse it and store in cache in ADAL, but never use. Please correct me if I missed something.

In case if *ADTokenCacheItem* has both AT&RT we map it to *MSIDToken* with type *MSIDTokenTypeRefreshToken* and we ignore value of AT. If we need to get MSIDToken with type *MSIDTokenTypeAccessToken*, RT in *ADTokenCacheItem* should be nil. Added tests that cover this logic.

### CODE REVIEW

File: IdentityCore/src/cache/MSIDUserInformation.m [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//  MSIDUserInformation.m
+//  IdentityCore
+//
+//  Created by Sergey Demchenko on 12/7/17.
+//  Copyright © 2017 Microsoft. All rights reserved.
+//
+
+#import "MSIDUserInformation.h"
+
+@implementation MSIDUserInformation
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDUserInformation.m]: MSIDUserInformation will need backward compatibility class name as well.

File: IdentityCore/src/cache/MSIDUserInformation.m [hunk 2]
```diff
@@ -0,0 +1,32 @@
+//
+//  MSIDUserInformation.m
+//  IdentityCore
+//
+//  Created by Sergey Demchenko on 12/7/17.
+//  Copyright © 2017 Microsoft. All rights reserved.
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDUserInformation.m]: please use the default header :)

File: IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m [hunk 3]
```diff
@@ -23,17 +23,29 @@
...
-    return nil;
+    return [NSKeyedArchiver archivedDataWithRootObject:token];
 }
 
 - (MSIDToken *)deserialize:(NSData *)data
 {
-    return nil;
+    NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:data];
+    [unarchiver setClass:MSIDUserInformation.class forClassName:@"ADUserInformation"];
+    MSIDToken *token = [unarchiver decodeObjectOfClass:MSIDToken.class forKey:NSKeyedArchiveRootObjectKey];
```
Comment (oldalton) [inline IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m]: why is this necessary? Isn't NSKeyedUnarchiver setClass:forClassName: enough?

File: IdentityCore/src/cache/MSIDUserInformation.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDUserInformation.m]: done

File: IdentityCore/src/cache/MSIDUserInformation.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDUserInformation.m]: It is specified in MSIDKeyedArchiverSerializer.m

File: IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m]: If we set class name globally in "NSKeyedUnarchiver setClass:forClassName:" it will conflict with archiving ADUserInformation in ADAL. Here we do manipulation with name locally so it will not affect other archivers.

File: IdentityCore/src/cache/MSIDUserInformation.m (hunk 1 above)
Comment (oldalton) [reply IdentityCore/src/cache/MSIDUserInformation.m]: I see, but do you need to set it each time separately or you could set it globally?

File: IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m (hunk 3 above)
Comment (oldalton) [reply IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m]: I see, but do we still want to archive ADUserInformation in ADAL?

Comment (antrix1989) [reply IdentityCore/src/cache/serializers/MSIDKeyedArchiverSerializer.m]: We will not need to archive ADUserInformation anymore, but I still think it is safer to do name manipulation locally. What advantages will we get from doing it globally?

File: IdentityCore/src/cache/MSIDToken.h [hunk 4]
```diff
@@ -23,22 +23,25 @@
...
 @property (readonly) NSDate *expiresOn;
 @property (readonly) NSString *familyId;
 @property (readonly) NSDictionary *clientInfo;
 @property (readonly) NSDictionary *additionalServerInfo;
-
 @property (readonly) MSIDTokenType tokenType;
+@property (readonly) NSString *resource;
+@property (readonly) NSString *authority;
+@property (readonly) NSString *clientId;
+
```
Comment (unpluggedk) [inline IdentityCore/src/cache/MSIDToken.h]: if we are to add like this, I suggest we also add scopes here too
@property (readonly) NSOrderedSet<NSString *> *scopes;

File: IdentityCore/src/cache/MSIDToken.h [hunk 5]
```diff
@@ -23,22 +23,25 @@
...
 
-@interface MSIDToken : NSObject
+@interface MSIDToken : NSObject <NSSecureCoding>
 
 @property (readonly) NSString *token;
-
 @property (readonly) NSString *idToken;
 @property (readonly) NSDate *expiresOn;
 @property (readonly) NSString *familyId;
 @property (readonly) NSDictionary *clientInfo;
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDToken.h]: Question for future: do we actually need to store the whole clientInfo? I'd suggest storing just userIdentifier

File: IdentityCore/src/cache/MSIDToken.m [hunk 6]
```diff
@@ -22,7 +22,130 @@
...
+        _tokenType = MSIDTokenTypeRefreshToken;
+    }
+    else
+    {
+        _token = accessToken;
+        _tokenType = MSIDTokenTypeAccessToken;
+    }
+    
+    _additionalServerInfo = [coder decodeObjectOfClass:[NSDictionary class] forKey:@"additionalServer"];
+    _clientInfo = [coder decodeObjectOfClass:[NSMutableDictionary class] forKey:@"additionalClient"];
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDToken.m]: clientInfo is not additionalClient. additionalClient is just additional properties to store with the token and not used by ADAL. clientInfo is additional data returned by the STS that has unique user identifier.

File: IdentityCore/src/cache/MSIDToken.m [hunk 7]
```diff
@@ -22,7 +22,130 @@
...
+    if (self.tokenType == MSIDTokenTypeRefreshToken)
+    {
+        [coder encodeObject:_token forKey:@"refreshToken"];
+    }
+    else
+    {
+        [coder encodeObject:_token forKey:@"accessToken"];
+    }
+    
+    [coder encodeObject:_clientInfo forKey:@"additionalClient"];
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDToken.m]: same comment here

File: IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m [hunk 8]
```diff
@@ -0,0 +1,84 @@
+//
+//  MSIDKeyedArchiverSerializerTests.m
+//  IdentityCore
+//
+//  Created by Sergey Demchenko on 12/7/17.
+//  Copyright © 2017 Microsoft. All rights reserved.
```
Comment (oldalton) [inline IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m]: please update the header

File: IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    [expectedToken setValue:[NSDate new] forKey:@"expiresOn"];
+    [expectedToken setValue:@"familyId value" forKey:@"familyId"];
+    [expectedToken setValue:@{@"key" : @"value"} forKey:@"clientInfo"];
+    [expectedToken setValue:@{@"key2" : @"value2"} forKey:@"additionalServerInfo"];
+    
+    NSData *data = [serializer serialize:expectedToken];
+    MSIDToken *resultToken = [serializer deserialize:data];
+    
+    XCTAssertNotNil(data);
+    XCTAssertEqualObjects(resultToken, expectedToken);
```
Comment (oldalton) [inline IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m]: can we also move the token isEqual tests from ADAL?

File: IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m (hunk 8 above)
Comment (antrix1989) [reply IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m]: done

File: IdentityCore/src/cache/MSIDToken.h (hunk 5 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDToken.h]: "userIdentifier" is a combination of "uid" + "utid". If we store "userIdentifier" only, we should be ready to parse it back to "uid" & "utid".

Comment (oldalton) [reply IdentityCore/src/cache/MSIDToken.h]: The idea is that the client shouldn't try to parse or use those values other than for user identification. And user identification is uid.utid

File: IdentityCore/src/cache/MSIDToken.m (hunk 6 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDToken.m]: fixed

File: IdentityCore/src/cache/MSIDToken.m (hunk 7 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDToken.m]: fixed

File: IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m (hunk 9 above)
Comment (antrix1989) [reply IdentityCore/tests/MSIDKeyedArchiverSerializerTests.m]: done

File: IdentityCore/src/cache/MSIDToken.h (hunk 4 above)
Comment (antrix1989) [reply IdentityCore/src/cache/MSIDToken.h]: added

File: IdentityCore/src/cache/MSIDToken.m [hunk 10]
```diff
@@ -22,7 +22,134 @@
...
+        _tokenType = MSIDTokenTypeRefreshToken;
+    }
+    else
+    {
+        _token = accessToken;
+        _tokenType = MSIDTokenTypeAccessToken;
+    }
+    
+    _additionalServerInfo = [coder decodeObjectOfClass:[NSDictionary class] forKey:@"additionalServer"];
+    _clientInfo = [coder decodeObjectOfClass:[NSMutableDictionary class] forKey:@"clientInfo"];
```
Comment (oldalton) [inline IdentityCore/src/cache/MSIDToken.m]: why it's NSMutableDictionary?

Comment (antrix1989) [reply IdentityCore/src/cache/MSIDToken.m]: my mistake, fixed it
